src/main/java/org/deepsymmetry/beatlink/VirtualRekordbox.java

```diff
@@ -415,16 +421,38 @@ private DeviceUpdate buildUpdate(DatagramPacket packet) {
                         lastValidStatusFlagBytes.put(rawDeviceNumber, reportedStatusFlags);  // Record in case next packet for this device is malformed.
                     }
 
+                    // The raw ID sent from the Opus, we just need this to detect changes.
+                    // We can't use CdjStatus.getRekordboxId() because it still refers to the
+                    // playerToDeviceSqlRekordboxId map that hasn't been updated yet.
+                    int rawRekordboxId = (int)Util.bytesToNumber(packet.getData(), 0x2c, 4);
+
                     CdjStatus status = new CdjStatus(packet, hadToRecoverStatusFlags);
 
-                    // If source player number is zero the deck does not have a song loaded, clear the PSSI and source slot we had for that player.
-                    // This also represents a new track load.
-                    // We also need to set the DeviceSQL rekordbox ID to 0 for the player, so we don't
-                    // reflect the old track right when the new track is loaded (it takes a few seconds to match the PSSI).
-                    if (status.getTrackSourcePlayer() == 0) {
+                    // Get previous ID for comparison
+                    Integer previousId = previousRawRekordboxIds.get(status.getDeviceNumber());
+                    
+                    // Determine if track has changed and if it's currently loaded
+                    boolean isLoaded = status.getTrackSourcePlayer() != 0;
+                    boolean trackChanged = previousId != null && previousId != rawRekordboxId && rawRekordboxId != 0;
```

Comment:
I don’t think you want the `null` check here. The track has still changed even if the previous ID was `null`, hasn’t it?

Reply:
Even if this is to avoid null pointer exceptions, I think the logic is backwards. If the previous ID was `null`, then this should always be considered a change.